### ISSUE TITLE
org.elasticsearch.index.fielddata.plain.DoubleArrayIndexFieldData cannot be cast to org.elasticsearch.index.fielddata.IndexGeoPointFieldData

### ISSUE THREAD

**Issue by hero007asd** [elasticsearch/elasticsearch#9347]
hi,I use ES1.4.1 and  I met a problem, as follows:
> in my mappings 

```
"location": {
    "type": "geo_point",
    "fielddata": {
        "precision": "1mm",
        "format": "compressed"
    }
},
```
> after I search like this:

```
{
  "sort": {
    "_geo_distance": {
      "location": [
        119.95720242,
        29.15949412
      ],
      "order": "asc",
      "unit": "m",
      "mode": "avg",
      "distance_type": "sloppy_arc"
    }
  }
}
```

An exception out:
```
ClassCastException[org.elasticsearch.index.fielddata.plain.DoubleArrayIndexFieldData cannot be cast to org.elasticsearch.index.fielddata.IndexGeoPointFieldData
Caused by: java.lang.ClassCastException: org.elasticsearch.index.fielddata.plain.DoubleArrayIndexFieldData cannot be cast to org.elasticsearch.index.fielddata.IndexGeoPointFieldData
        at org.elasticsearch.search.sort.GeoDistanceSortParser.parse(GeoDistanceSortParser.java:139)
        at org.elasticsearch.search.sort.SortParseElement.addCompoundSortField(SortParseElement.java:144)
        at org.elasticsearch.search.sort.SortParseElement.parse(SortParseElement.java:96)
        at org.elasticsearch.search.SearchService.parseSource(SearchService.java:665)
```

**Comment by clintongormley** [comment]
Hi @hero007asd 

It looks like either:

* you have two fields called `location` in the same index, with different mappings
* you are relying on dynamic mapping 

Either way, one shard seems to think that that field is a double, while the other shards think that it is a geo-point.

Both of these issues will be fixed by #8870

**Comment by hero007asd** [comment]
Thank you, @clintongormley , it works for me!
